File: packages/python/trikhub/cli/templates/python.py
"""Python Trik Template Generator (v2).

Generates all files needed for a v2 Python trik project.
Supports conversational and tool modes using the trikhub SDK.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any


@dataclass
class PyTemplateConfig:
    name: str
    display_name: str
    description: str
    author_name: str
    author_github: str
    category: str
    enable_storage: bool
    enable_config: bool
    agent_mode: str  # 'conversational' | 'tool'
    handoff_description: str = ""
    domain_tags: list[str] = field(default_factory=list)
    tool_names: list[str] = field(default_factory=list)


@dataclass
class GeneratedFile:
    path: str
    content: str


def _to_package_name(name: str) -> str:
    return name.replace("-", "_")


def _to_pascal_case(name: str) -> str:
    return "".join(part.capitalize() for part in name.split("-"))


def generate_python_project(config: PyTemplateConfig) -> list[GeneratedFile]:
    files: list[GeneratedFile] = []

    files.append(GeneratedFile("manifest.json", _generate_manifest(config)))
    files.append(GeneratedFile("trikhub.json", _generate_trikhub_json(config)))
    files.append(GeneratedFile("pyproject.toml", _generate_pyproject(config)))
    files.append(GeneratedFile("test.py", _generate_test_py(config)))
    files.append(GeneratedFile(".gitignore", _generate_gitignore()))

    if config.agent_mode == "conversational":
        files.append(GeneratedFile("src/agent.py", _generate_conversational_agent(config)))
        files.append(GeneratedFile("src/tools/example.py", _generate_example_tool()))
        files.append(GeneratedFile("src/prompts/system.md", _generate_system_prompt(config)))
    else:
        files.append(GeneratedFile("src/agent.py", _generate_tool_agent(config)))

    # Include manifest inside src/ so it's included in the pip wheel.
    # The gateway can discover it after `pip install`.
    files.append(GeneratedFile("src/manifest.json", _generate_package_manifest(config)))

    return files


def _generate_manifest(config: PyTemplateConfig) -> str:
    agent: dict[str, Any] = {
        "mode": config.agent_mode,
        "domain": config.domain_tags or [config.category],
    }

    if config.agent_mode == "conversational":
        agent["handoffDescription"] = config.handoff_description or config.description
        agent["systemPromptFile"] = "./src/prompts/system.md"
        agent["model"] = {"capabilities": ["tool_use"]}

    manifest: dict[str, Any] = {
        "schemaVersion": 2,
        "id": config.name,
        "name": config.display_name,
        "description": config.description,
        "version": "0.1.0",
        "agent": agent,
        "limits": {"maxTurnTimeMs": 30000},
        "entry": {
            "module": "./src/agent.py",
            "export": "agent",
            "runtime": "python",
        },
        "author": f"@{config.author_github}",
        "repository": f"https://github.com/{config.author_github}/{config.name}",
    }

    if config.agent_mode == "tool":
        tools: dict[str, Any] = {}
        for tool_name in (config.tool_names or ["exampleTool"]):
            tools[tool_name] = {
                "description": f"TODO: describe {tool_name}",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                        "query": {"type": "string", "description": "Input query"},
                    },
                    "required": ["query"],
                },
                "outputSchema": {
                    "type": "object",
                    "properties": {
                        "result": {
                            "type": "string",
                            "maxLength": 1000,
                            "pattern": "^.{0,1000}$",
                        },
                    },
                },
                "outputTemplate": "Result: {{result}}",
            }
        manifest["tools"] = tools
    else:
        manifest["tools"] = {
            "exampleTool": {
                "description": "An example tool",
            },
        }

    if config.enable_storage:
        manifest["capabilities"] = {
            "storage": {"enabled": True, "maxSizeBytes": 1048576, "persistent": True},
        }

    if config.enable_config:
        manifest["config"] = {
            "required": [{"key": "API_KEY", "description": "Your API key"}],
            "optional": [],
        }

    return json.dumps(manifest, indent=2) + "\n"


def _generate_package_manifest(config: PyTemplateConfig) -> str:
    """Generate a manifest for inclusion inside the pip wheel.

    Paths are adjusted to be relative to the src/ directory instead of the repo root.
    """
    manifest = json.loads(_generate_manifest(config))

    # Adjust entry.module to be relative to the package directory
    manifest["entry"]["module"] = "./agent.py"

    # Adjust systemPromptFile if present
    if "systemPromptFile" in manifest.get("agent", {}):
        manifest["agent"]["systemPromptFile"] = "./prompts/system.md"

    return json.dumps(manifest, indent=2) + "\n"


def _generate_trikhub_json(config: PyTemplateConfig) -> str:
    trikhub = {
        "displayName": config.display_name,
        "shortDescription": config.description,
        "categories": [config.category],
        "keywords": [config.name],
        "author": {
            "name": config.author_name,
            "github": config.author_github,
        },
        "repository": f"https://github.com/{config.author_github}/{config.name}",
    }
    return json.dumps(trikhub, indent=2) + "\n"


def _generate_pyproject(config: PyTemplateConfig) -> str:
    deps = [
        '"trikhub>=0.6.0"',
    ]
    if config.agent_mode == "conversational":
        deps.extend([
            '"langchain-anthropic>=0.3.0"',
            '"langgraph>=0.2.0"',
        ])

    deps_str = ",\n    ".join(deps)

    return f'''[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "{config.name}"
version = "0.1.0"
description = "{config.description}"
readme = "README.md"
requires-python = ">=3.10"
license = {{text = "MIT"}}
authors = [
    {{ name = "{config.author_name}" }}
]
dependencies = [
    {deps_str},
]

[project.urls]
Repository = "https://github.com/{config.author_github}/{config.name}"

[tool.hatch.build.targets.wheel]
packages = ["src"]

[tool.hatch.build.targets.wheel.force-include]
"manifest.json" = "src/manifest.json"

[tool.hatch.build.targets.sdist]
include = [
    "src/**",
    "manifest.json",
    "trikhub.json",
    "README.md",
]
'''


def _generate_conversational_agent(config: PyTemplateConfig) -> str:
    class_name = _to_pascal_case(config.name)
    return f'''"""
{config.display_name} — conversational agent

Uses LangGraph ReAct pattern with the TrikHub SDK.
"""

from __future__ import annotations

from langchain_anthropic import ChatAnthropic
from langgraph.prebuilt import create_react_agent

from trikhub.sdk import wrap_agent, transfer_back_tool
from trikhub.manifest import TrikContext

from .tools.example import example_tool


def create_agent(context: TrikContext):
    """Factory that creates a new agent per session."""
    model = ChatAnthropic(model="claude-sonnet-4-20250514")

    # Read system prompt from file (loaded by gateway via manifest.agent.systemPromptFile)
    tools = [example_tool, transfer_back_tool]

    return create_react_agent(
        model,
        tools=tools,
    )


# Export wrapped agent for the TrikHub worker
agent = wrap_agent(create_agent)
'''


def _generate_tool_agent(config: PyTemplateConfig) -> str:
    handlers: list[str] = []
    for tool_name in (config.tool_names or ["exampleTool"]):
        handlers.append(f'''    "{tool_name}": {_to_handler_name(tool_name)},''')

    func_defs: list[str] = []
    for tool_name in (config.tool_names or ["exampleTool"]):
        func_name = _to_handler_name(tool_name)
        func_defs.append(f'''
async def {func_name}(input_data: dict, context) -> dict:
    """Handle {tool_name} tool calls."""
    query = input_data.get("query", "")
    return {{"result": f"Processed: {{query}}"}}
''')

    handlers_str = "\n".join(handlers)
    func_defs_str = "\n".join(func_defs)

    return f'''"""
{config.display_name} — tool mode agent

Exports native tools via wrapToolHandlers from the TrikHub SDK.
"""

from __future__ import annotations

from trikhub.sdk import wrap_tool_handlers

{func_defs_str}

# Export wrapped tool handlers for the TrikHub worker
agent = wrap_tool_handlers({{
{handlers_str}
}})
'''


def _to_handler_name(tool_name: str) -> str:
    """Convert camelCase tool name to snake_case handler name."""
    result = []
    for i, char in enumerate(tool_name):
        if char.isupper() and i > 0:
            result.append("_")
        result.append(char.lower())
    return "handle_" + "".join(result)


def _generate_example_tool() -> str:
    return '''"""Example tool for the conversational agent."""

from langchain_core.tools import tool


@tool
def example_tool(query: str) -> str:
    """Search for information about a topic.

    Args:
        query: The search query.
    """
    return f"Result for: {query}"
'''


def _generate_system_prompt(config: PyTemplateConfig) -> str:
    return f"""You are {config.display_name}, a helpful assistant.

{config.description}

When you have finished helping the user, use the transfer_back tool to return control to the main agent.
"""


def _generate_test_py(config: PyTemplateConfig) -> str:
    if config.agent_mode == "tool":
        tool_name = (config.tool_names or ["exampleTool"])[0]
        handler_name = _to_handler_name(tool_name)
        return f'''"""Local test script — run with: python test.py"""

import asyncio
from src.agent import agent


async def main():
    result = await agent.execute_tool(
        "{tool_name}",
        {{"query": "hello world"}},
        context=None,
    )
    print(result)


if __name__ == "__main__":
    asyncio.run(main())
'''
    else:
        return f'''"""Local test script — run with: python test.py"""

import asyncio
from src.agent import agent


async def main():
    result = await agent.process_message(
        "Hello! What can you do?",
        session_id="test-session",
        context=None,
    )
    print(result)


if __name__ == "__main__":
    asyncio.run(main())
'''


def _generate_gitignore() -> str:
    return """__pycache__/
*.py[cod]
*$py.class
*.so
.Python
build/
dist/
*.egg-info/
.eggs/
*.egg
.venv/
venv/
.DS_Store
"""

File: packages/js/cli/src/templates/python.ts
/**
 * v2 Python scaffold template.
 *
 * Generates a complete v2 trik project structure for Python,
 * mirroring the TypeScript template but using Python conventions.
 */

import type { InitConfig } from './typescript.js';

// ============================================================================
// Helpers
// ============================================================================

/** Convert trik name (dashes) to Python package name (underscores) */
function toPythonPackage(name: string): string {
  return name.replace(/-/g, '_');
}

// ============================================================================
// File generators
// ============================================================================

function generateManifest(config: InitConfig): string {
  const isToolMode = config.agentMode === 'tool';
  const pkg = toPythonPackage(config.name);

  const agent: Record<string, unknown> = {
    mode: config.agentMode,
    domain: config.domainTags,
  };

  if (!isToolMode) {
    agent.handoffDescription = config.handoffDescription;
    agent.systemPromptFile = `./src/${pkg}/prompts/system.md`;
    agent.model = { capabilities: ['tool_use'] };
  }

  const manifest: Record<string, unknown> = {
    schemaVersion: 2,
    id: config.name,
    name: config.displayName,
    description: config.description,
    version: '0.1.0',
    agent,
  };

  // Tools block
  if (isToolMode && config.toolNames.length > 0) {
    const tools: Record<string, Record<string, unknown>> = {};
    for (const toolName of config.toolNames) {
      tools[toolName] = {
        description: `TODO: describe ${toolName}`,
        inputSchema: {
          type: 'object',
          properties: {
            query: { type: 'string', maxLength: 200 },
          },
          required: ['query'],
        },
        outputSchema: {
          type: 'object',
          properties: {
            status: { type: 'string', enum: ['success', 'error'] },
            resultId: { type: 'string', format: 'id' },
          },
          required: ['status'],
        },
        outputTemplate: `${toolName}: {{status}} ({{resultId}})`,
      };
    }
    manifest.tools = tools;
  } else {
    manifest.tools = {
      exampleTool: {
        description: 'An example tool',
      },
    };
  }

  if (config.enableStorage) {
    manifest.capabilities = {
      storage: { enabled: true },
    };
  }

  manifest.limits = { maxTurnTimeMs: 30000 };
  manifest.entry = {
    module: `./src/${pkg}/main.py`,
    export: 'default',
    runtime: 'python',
  };
  manifest.author = config.authorName;

  if (config.enableConfig) {
    manifest.config = {
      optional: [
        { key: 'ANTHROPIC_API_KEY', description: 'Anthropic API key for the agent' },
      ],
    };
  }

  return JSON.stringify(manifest, null, 2);
}

function generateTrikhubJson(config: InitConfig): string {
  const metadata = {
    displayName: config.displayName,
    shortDescription: config.description,
    categories: [config.category],
    keywords: [] as string[],
    author: {
      name: config.authorName,
      github: config.authorGithub,
    },
    repository: `https://github.com/${config.authorGithub}/${config.name}`,
  };

  return JSON.stringify(metadata, null, 2);
}

function generatePyprojectToml(config: InitConfig): string {
  const isToolMode = config.agentMode === 'tool';
  const pkg = toPythonPackage(config.name);

  const deps = ['    "trikhub-sdk>=0.1.0",'];
  if (!isToolMode) {
    deps.push('    "langchain-anthropic>=0.3.0",');
    deps.push('    "langchain-core>=0.3.0",');
    deps.push('    "langgraph>=0.2.0",');
  }

  return `[build-system]
requires = ["setuptools>=61.0", "wheel"]
build-backend = "setuptools.build_meta"

[project]
name = "${config.name}"
version = "0.1.0"
description = "${config.description}"
requires-python = ">=3.10"
license = "MIT"

dependencies = [
${deps.join('\n')}
]

[tool.setuptools.packages.find]
where = ["src"]
include = ["${pkg}*"]
`;
}

function generateToolModeMain(config: InitConfig): string {
  const handlers = config.toolNames.map((name) =>
    `async def ${toSnakeCase(name)}(input: dict[str, Any], context: TrikContext) -> dict[str, Any]:
    # TODO: Implement ${name}
    return {"result": "Not implemented"}`
  ).join('\n\n\n');

  const handlerMap = config.toolNames.map((name) =>
    `    "${name}": ${toSnakeCase(name)},`
  ).join('\n');

  return `"""
${config.displayName} — tool-mode trik.

Exports native tools to the main agent. No handoff, no session.
Uses wrap_tool_handlers() for native tool export.
"""

from typing import Any

from trikhub.sdk import wrap_tool_handlers, TrikContext


${handlers}


default = wrap_tool_handlers({
${handlerMap}
})
`;
}

function generateConversationalMain(config: InitConfig): string {
  const apiKeyAccess = config.enableConfig
    ? 'config.get("ANTHROPIC_API_KEY")'
    : 'os.environ.get("ANTHROPIC_API_KEY")';

  return `"""
${config.displayName} — conversational trik.

Uses the wrap_agent() pattern for multi-turn conversation via handoff.
"""

from __future__ import annotations

import os
from pathlib import Path

from langchain_anthropic import ChatAnthropic
from langchain_core.tools import tool
from langgraph.prebuilt import create_react_agent

from trikhub.sdk import wrap_agent, transfer_back_tool, TrikContext


_PROMPT_PATH = Path(__file__).parent / "prompts" / "system.md"
_SYSTEM_PROMPT = _PROMPT_PATH.read_text(encoding="utf-8")


@tool
async def example_tool(query: str) -> str:
    """An example tool — replace with your own implementation."""
    return f"Processed: {query}"


default = wrap_agent(lambda context: create_react_agent(
    model=ChatAnthropic(
        model="claude-sonnet-4-20250514",
        api_key=${apiKeyAccess},
    ),
    tools=[example_tool, transfer_back_tool],
    prompt=_SYSTEM_PROMPT,
))
`;
}

function generateMainPy(config: InitConfig): string {
  if (config.agentMode === 'tool') {
    return generateToolModeMain(config);
  }
  return generateConversationalMain(config);
}

function generateSystemPrompt(config: InitConfig): string {
  const domainStr = config.domainTags.join(', ');

  return `# ${config.displayName}

You are ${config.displayName}, a specialized assistant for ${config.description.toLowerCase()}.

## Your capabilities
- **example_tool**: An example tool

## Guidelines
- Focus on tasks within your domain: ${domainStr}
- When the user's request is outside your expertise, use the transfer_back tool
- Provide clear, actionable responses

## Transfer back
Use the \`transfer_back\` tool when:
- The user's request is outside your domain
- You've completed the task and the user wants to do something else
- The user explicitly asks to go back
`;
}

function generateGitignore(): string {
  return `__pycache__/
*.egg-info/
*.pyc
dist/
build/
*.egg
.venv/
.env
.trikhub/secrets.json
`;
}

function generateReadme(config: InitConfig): string {
  const domainStr = config.domainTags.join(', ');
  const isToolMode = config.agentMode === 'tool';
  const pkg = toPythonPackage(config.name);

  const devSection = isToolMode
    ? `- Implement your tool handlers in \`src/${pkg}/main.py\`
- Update inputSchema/outputSchema in \`manifest.json\``
    : `- Edit your agent logic in \`src/${pkg}/main.py\`
- Add tools as \`@tool\` decorated functions
- Customize the system prompt in \`src/${pkg}/prompts/system.md\``;

  const archSection = isToolMode
    ? `Tools from this trik appear as native tools on the main agent — no handoff, no session.`
    : `The main agent routes conversations to this trik using a \`talk_to_${config.name}\` tool.
When done, use the \`transfer_back\` tool to return control.`;

  return `# ${config.displayName}

${config.description}

## Getting Started

1. Create a virtual environment: \`python -m venv .venv && source .venv/bin/activate\`
2. Install dependencies: \`pip install -e .\`
3. Validate: \`trik lint .\`
4. Publish: \`trik publish\`

## Development

${devSection}

## Architecture

This trik uses the TrikHub v2 architecture:
- **Mode**: ${config.agentMode}
- **Domain**: ${domainStr}

${archSection}
`;
}

/** Convert camelCase to snake_case */
function toSnakeCase(str: string): string {
  return str.replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`);
}

// ============================================================================
// Public API
// ============================================================================

/**
 * Generate a complete v2 Python trik project.
 *
 * @returns Map of { relativePath: fileContent } for all project files.
 */
export function generatePythonProject(config: InitConfig): Record<string, string> {
  const files: Record<string, string> = {};
  const pkg = toPythonPackage(config.name);

  // Core config files (at project root)
  files['manifest.json'] = generateManifest(config);
  files['trikhub.json'] = generateTrikhubJson(config);
  files['pyproject.toml'] = generatePyprojectToml(config);
  files['.gitignore'] = generateGitignore();
  files['README.md'] = generateReadme(config);

  // Python package files
  files[`src/${pkg}/__init__.py`] = '';
  files[`src/${pkg}/main.py`] = generateMainPy(config);

  if (config.agentMode === 'conversational') {
    files[`src/${pkg}/prompts/system.md`] = generateSystemPrompt(config);
  }

  return files;
}
